component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java

```diff
@@ -161,58 +160,107 @@ protected void initiateAuthenticationRequest(HttpServletRequest request,
+                        break;
+                    }
                 }
-                email = getEmailValueForUsername(username, context);
-                if (StringUtils.isEmpty(email)) {
-                    if (request.getParameter(EmailOTPAuthenticatorConstants.EMAIL_ADDRESS) == null) {
-                        redirectToEmailAddressReqPage(response, context, emailOTPParameters, queryParams, username);
-                    } else {
-                        updateEmailAddressForUsername(context, request, username);
-                        email = getEmailValueForUsername(username, context);
+
+                if (username == null) {
+                    if (log.isDebugEnabled()) {
+                        log.debug("Cannot find the subject attributed step with authenticated user.");
                     }
+                    throw new AuthenticationFailedException
+                            ("Authentication failed. Cannot find the subject attributed step with authenticated user.");
                 }
-                if (StringUtils.isNotEmpty(email)) {
-                    processEmailOTPFlow(request, response, email, username, queryParams, context);
+
+                if (isLocalUser) {
+                    handleEmailOTPForLocalUser(username, authenticatedUser, context, emailOTPParameters,
+                            isEmailOTPMandatory, queryParams, request, response);
+
+                } else {
+                    handleEmailOTPForFederatedUser(sendOtpToFederatedEmail, isEmailOTPMandatory, context,
+                            userAttributes, federatedEmailAttributeKey, authenticatedUser, username, emailOTPParameters,
+                            queryParams, request, response);
                 }
+
             } else {
-                if (log.isDebugEnabled()) {
-                    log.debug("Process with the first step (basic) authenticator only");
+                // This block need to be revised and recommended to be removed
+                FederatedAuthenticatorUtil.setUsernameFromFirstStep(context);
+                String username = String.valueOf(context.getProperty(EmailOTPAuthenticatorConstants.USER_NAME));
+                authenticatedUser = (AuthenticatedUser) context.getProperty
+                        (EmailOTPAuthenticatorConstants.AUTHENTICATED_USER);
+                // find the authenticated user.
+                if (authenticatedUser == null) {
+                    if (log.isDebugEnabled()) {
+                        log.debug("Cannot find the authenticated user, the username : " + username + " may be null");
+                    }
+                    throw new AuthenticationFailedException
+                            ("Authentication failed!. Cannot find the authenticated user, the username : "
+                                    + username + " may be null");
+                }
+
+                boolean isUserExistence = FederatedAuthenticatorUtil.isUserExistInUserStore(username);
```

Comment:
Are we still going with this check?

Reply:
We haven't remove the previous code. If the usecase attribute is configured, then this check will be executed. Otherwise, the fixed flow will be executed.